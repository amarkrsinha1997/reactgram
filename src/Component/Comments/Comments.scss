.comments {
  background: white;
  width: 100%;
}

.comment {
  padding: 5px 10px;
  background: #eeeeee;
  color: #333333;
  margin: 2.5px 0;
  border-radius: 3px;
  display: flex;
  justify-content: space-between;
}

.comment-message {
  align-self: flex-start;
  flex: 1;
}

.user {
  font-weight: 700;
  margin-right: 5px;
}

.icon {
  align-self: flex-end;
  margin-left: 5px;
  cursor: pointer;
}

.input-box {
  border: none;
  padding: 5px;
  flex: 1;
  border-radius: 3px;
}