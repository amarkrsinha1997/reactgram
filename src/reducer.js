import { PostModel } from "./Model/PostModel";

const actionTypes = {
  ADD_COMMENT: "ADD_COMMENT",
  DELETE_COMMENT: "DELETE_COMMENT",
  EDIT_COMMENT: "EDIT_COMMENT",
  LIKE_COMMENT: "LIKE_COMMENT"
}
const reducer = (state, action) => {
  const { posts } = state;
  const { postId, commentId, comment } = action.payload;
  switch (action.type) {
    case actionTypes.ADD_COMMENT:
      return {
        ...state,
        posts: PostModel.addComment(posts, postId, comment),
      };
    case actionTypes.DELETE_COMMENT:
      return {
        ...state,
        posts: PostModel.removeComment(posts, postId, commentId),
      }
    case actionTypes.EDIT_COMMENT:
      return {
        ...state,
        posts: PostModel.updateComment(posts, postId, commentId, comment),
      };
    case actionTypes.LIKE_COMMENT: 
      return {
        ...state,
        posts: PostModel.like(posts, postId)
      }
    default:
      return state;
  }
};

export { reducer, actionTypes}