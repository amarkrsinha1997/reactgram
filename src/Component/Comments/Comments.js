import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './Comments.scss';
import { commentPropTypes } from '../../reusablePropType';
import { actionTypes } from '../../reducer';

const editIcon = "https://img.icons8.com/cute-clipart/30/000000/edit.png"
const cancelIcon = "https://img.icons8.com/cute-clipart/30/000000/cancel.png";

const Comment = (props) => {
  const { comment, handleChange } = props;
  const [commentInput, setCommentInput] = useState(comment.text);
  const [editable, setEditable] = useState(false);

  const handleIconClick = () => setEditable((prevEditable) => !prevEditable);

  const handleCommentEdit = async (event) => {
    const { value: updatedCommentMessage } = event.target;
    setCommentInput(updatedCommentMessage);
  }

  const handleSubmit = async event => {
    if(event.key === 'Enter') {
      await handleChange(comment.id, commentInput);
      handleIconClick();
    }
  }
  
  return (
    <React.Fragment>
      <div className="comment">
        <span className="user">{comment.user}: </span>
        {
          !editable ?
            <span className="comment-message">
              {commentInput}
            </span> :
            <input 
              data-testid="edit-input"
              type="text" 
              onChange={handleCommentEdit}
              onKeyDown={handleSubmit} 
              name="edit-comment" 
              className="input-box" 
              value={commentInput} 
              />
        }
        {
          !editable ?
            <img
              src={editIcon}
              alt="edit-icon"
              className="icon"
              onClick={handleIconClick} /> :
            <img
              src={cancelIcon}
              alt="cancel-edit-icon"
              className="icon"
              onClick={handleIconClick} />
        }
      </div>
    </React.Fragment>
  )
}

Comment.propTypes = {
  comment: commentPropTypes,
  handleChange: PropTypes.func.isRequired,
}

const Comments = (props) => {
  const { postId, code, comments, dispatch } = props;

  const handleChange = (postId) => async (commentId, comment) => {
    await dispatch({
      type: actionTypes.EDIT_COMMENT,
      payload: { postId, commentId, comment }
    });
  }


  return (
    <div className="comments" data-testid={code}>
      {comments.map((comment) =>
        <Comment key={code + comment.id} comment={comment} handleChange={handleChange(postId)} />
      )}
    </div>
  )
}

Comments.propTypes = {
  comments: PropTypes.arrayOf(commentPropTypes),
  dispatch: PropTypes.func.isRequired,
  postId: PropTypes.string.isRequired,
  code: PropTypes.string.isRequired
}
Comments.defaultProps = {
  comments: []
}
export default Comments

