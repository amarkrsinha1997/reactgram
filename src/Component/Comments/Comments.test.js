import React from "react";
import { render, fireEvent, waitForElementToBeRemoved,  } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Comments from "./Comments";
import { actionTypes } from "../../reducer";
describe('Comments', () => {
	let props;
	beforeEach(() => {
		props = {
			dispatch: jest.fn(),
			comments: [{id: 1,  user: 'Amar', text: 'nice dp!' }, {id: 2, user: 'Depak', text: 'Cool!'}],
			postId: "1234",
			code: 'AXU_Y21A_T'
		}
	})
	test("should render properly", () => {
		const wrapper = render(<Comments {...props}/>);
		expect(wrapper.container).toMatchSnapshot()
		expect(wrapper.getAllByAltText(/edit-icon/)).toHaveLength(2);
	})
	test("should show cancel item on click of edit button", () => {
		const wrapper = render(<Comments {...props} comments={[{ user: 'Amar', text: 'nice dp!' }]}/>);
		fireEvent.click(wrapper.getByAltText(/edit-icon/));
		expect(wrapper.getByAltText(/cancel-edit-icon/)).toBeInTheDocument();
	});
	test("should show input box when edit button is clicked.", () => {
		const wrapper = render(<Comments {...props} comments={[{ user: 'Amar', text: 'nice dp!' }]} />);
		fireEvent.click(wrapper.getByAltText(/edit-icon/));
		expect(wrapper.getByDisplayValue(/nice dp!/)).toBeInTheDocument();
	})
	test('should hide input box when edit button was clicked twice', () => {
		const wrapper = render(<Comments {...props} comments={[{ user: 'Amar', text: 'nice dp!' }]} />);
		fireEvent.click(wrapper.getByAltText(/edit-icon/));
		fireEvent.click(wrapper.getByAltText(/cancel-edit-icon/));
		expect(wrapper.queryByDisplayValue(/nice dp!/)).not.toBeInTheDocument();
	});

	test('should be able to change input value', async () => {
		const wrapper = render(<Comments {...props} comments={[{ user: 'Amar', text: 'nice dp!' }]} />);
		fireEvent.click(wrapper.getByAltText(/edit-icon/));
		userEvent.type(wrapper.getByTestId(/edit-input/), 'Hello');
		fireEvent.keyDown(wrapper.getByTestId(/edit-input/), { key: 'Enter' });
		
		await waitForElementToBeRemoved(() => wrapper.getByTestId(/edit-input/));

		expect(wrapper.queryByTestId(/edit-input/)).not.toBeInTheDocument();
		expect(props.dispatch).toHaveBeenCalledWith({
			type: actionTypes.EDIT_COMMENT,
			payload: { postId: props.postId, commentId: props.id, comment: 'Hello' }
		})
	})
	test('should focus on input when edit-icon clicked', () => {
		const wrapper = render(<Comments {...props} comments={[{ user: 'Amar', text: 'nice dp!' }]} />);
		fireEvent.click(wrapper.getByAltText(/edit-icon/));
		expect(wrapper.getByTestId(/edit-input/)).toHaveFocus();
	})
});