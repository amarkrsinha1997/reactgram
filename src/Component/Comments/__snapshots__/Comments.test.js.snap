// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments should render properly 1`] = `
<div>
  <div
    class="comments"
    data-testid="AXU_Y21A_T"
  >
    <div
      class="comment"
    >
      <span
        class="user"
      >
        Amar
        : 
      </span>
      <span
        class="comment-message"
      >
        nice dp!
      </span>
      <img
        alt="edit-icon"
        class="icon"
        src="https://img.icons8.com/cute-clipart/30/000000/edit.png"
      />
    </div>
    <div
      class="comment"
    >
      <span
        class="user"
      >
        Depak
        : 
      </span>
      <span
        class="comment-message"
      >
        Cool!
      </span>
      <img
        alt="edit-icon"
        class="icon"
        src="https://img.icons8.com/cute-clipart/30/000000/edit.png"
      />
    </div>
  </div>
</div>
`;
